src/google/protobuf/compiler/csharp/csharp_source_generator_base.cc

```diff
@@ -60,7 +60,7 @@ void SourceGeneratorBase::WriteGeneratedCodeAttributes(io::Printer* printer) {
 }
 
 std::string SourceGeneratorBase::class_access_level() {
-  return IsDescriptorProto(descriptor_) ? "internal" : "public";  // public_classes is always on.
+  return (IsDescriptorProto(descriptor_) || this->options()->internal_access) ? "internal": "public";
```

Comment:
Tiny nit if you have the energy: add a space before the colon.
(If not, I'll fix it up as part of another PR.)